[Video][Stacks] Fix folder stacks.

78andyp opened the pull request:
## Description

Folder stacks are useful for those who have ISOs or DVD/Bluray folders.
They are currently broken.

There were several issues:
- the code to generate folder stacks (`stack://`) was incomplete.
- folder stacks were then not recognised at various points in the play pathway (initially, transitioning from one item to another etc..)
- playing a stack from the main menu (not movie menu) was also broken (fixed in #24820)
- correct length of combined folder stack and streamdetails not saved correctly and not shown in bottom right
- routine to determine potential NFO/art path was broken

Also fixes:
- if you stop a DVD/Bluray on a menu after watching the film the menu length and not the film length is recoded (and shown in bottom right).
- one of the regexs for file stacks was to relaxed (see discussion in comments)

There a few differences now between file and folder stacks.
- For DVDs - each item is played in isolation, sequentially - you cannot move back and forth like you can with a file stack.
- For Blurays, now #24720 is in, bluray stacks will be built dynamically the first time they are played - and you will be able to move backwards. Once every item is played (and converted to bluray://) then a stack will behave exactly like a file stack.

## Motivation and context

Fixes #16109.

## How has this been tested?

Only locally currently.
With DVD/ISO/Bluray folders.
Also checked file stacks and individual DVD/Bluray/ISOs still play.

## What is the effect on users?

As above.

## Screenshots (if appropriate):

## Types of change
<!--- What type of change does your code introduce? Put an `x` with no space in all the boxes that apply like this: [X] -->
- [X] **Bug fix** (non-breaking change which fixes an issue)
- [ ] **Clean up** (non-breaking change which removes non-working, unmaintained functionality)
- [X] **Improvement** (non-breaking change which improves existing functionality)
- [ ] **New feature** (non-breaking change which adds functionality)
- [ ] **Breaking change** (fix or feature that will cause existing functionality to change)
- [ ] **Cosmetic change** (non-breaking change that doesn't touch code)
- [ ] **Student submission** (PR was done for educational purposes and will be treated as such)
- [ ] **None of the above** (please explain below)

## Checklist:
<!--- Go over all the following points, and put an `X` with no space in all the boxes that apply like this: [X] -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [ ] My code follows the **[Code Guidelines](https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md)** of this project 
- [ ] My change requires a change to the documentation, either Doxygen or wiki
- [ ] I have updated the documentation accordingly
- [ ] I have read the **[Contributing](https://github.com/xbmc/xbmc/blob/master/docs/CONTRIBUTING.md)** document
- [ ] I have added tests to cover my change
- [ ] All new and existing tests passed

ksooo commented on xbmc/cores/VideoPlayer/VideoPlayer.cpp:
```diff
@@ -8,6 +8,7 @@
 
 #include "VideoPlayer.h"
 
+#include "application/ApplicationStackHelper.h"
```
Making VideoPlayer dependent on Application is a nogo.

78andyp replied:
I've moved it - are you happy with where it is now (in ApplicationPlayerCallback)??

ksooo replied:
Yes, thanks.

ksooo commented on xbmc/application/Application.cpp:
```diff
@@ -2384,6 +2384,10 @@ bool CApplication::PlayFile(CFileItem item, const std::string& player, bool bRes
 
+  // Needed so stacks can be played properly from the main screen
+  if (URIUtils::IsStack(item.GetDynPath()))
+    item.SetPath(item.GetDynPath());
```
Although we do this at several places (mostly for historical reasons), every manipulation of fileitem's path after construction has to be avoided, as the path is the identity of the item. Changing it can have unpredictable side effects.

The dynpath is exactly there to hold the url of what needs to be played. Instead of manipulating the path, the code that handles the playback needs to be fixed to use item's dynpath, not the path.

78andyp replied:
Changed as suggested.

ksooo commented on xbmc/utils/URIUtils.cpp:
```diff
@@ -777,6 +777,24 @@ bool URIUtils::IsDVD(const std::string& strFile)
+{
+  std::string strFileLow = strFile;
+  StringUtils::ToLower(strFileLow);
+  if (strFileLow.find("video_ts.ifo") != std::string::npos)
+    return true;
+  return false;
```
`return (fileLow.find("video_ts.ifo") != std::string::npos));`

ksooo replied:
The implementation is not correct, as it will also return true if we pass in something containing "video_ts.ifo" in the middle of the path, like "/foo/video_ts.ifo/bar/baz.txt". It should look whether the passed in file ends (!) with "/video_ts.ifo" or if it does not start with a "/", equals "video_ts.ifo".

78andyp replied:
Done

ksooo commented on xbmc/utils/URIUtils.cpp:
```diff
@@ -777,6 +777,24 @@ bool URIUtils::IsDVD(const std::string& strFile)
   return false;
 }
 
+bool URIUtils::IsDVDFile(const std::string& strFile)
```
Remove `str`prefix from param name. Coding guidelines.

78andyp replied:
Done

ksooo commented on xbmc/utils/URIUtils.cpp:
```diff
@@ -777,6 +777,24 @@ bool URIUtils::IsDVD(const std::string& strFile)
 
+bool URIUtils::IsDVDFile(const std::string& strFile)
+{
+  std::string strFileLow = strFile;
```
Remove `str`prefix from var name. Coding guidelines.

78andyp replied:
Done

ksooo commented on xbmc/utils/URIUtils.cpp:
```diff
@@ -777,6 +777,24 @@ bool URIUtils::IsDVD(const std::string& strFile)
+  return false;
+}
+
+bool URIUtils::IsBDFile(const std::string& strFile)
```
Remove `str`prefix from param name. Coding guidelines.

78andyp replied:
Done

ksooo commented on xbmc/utils/URIUtils.cpp:
```diff
@@ -777,6 +777,24 @@ bool URIUtils::IsDVD(const std::string& strFile)
+
+bool URIUtils::IsBDFile(const std::string& strFile)
+{
+  std::string strFileLow = strFile;
```
Remove `str`prefix from var name. Coding guidelines.

78andyp replied:
Done

ksooo commented on xbmc/utils/URIUtils.cpp:
```diff
@@ -777,6 +777,24 @@ bool URIUtils::IsDVD(const std::string& strFile)
+{
+  std::string strFileLow = strFile;
+  StringUtils::ToLower(strFileLow);
+  if (strFileLow.find("index.bdmv") != std::string::npos)
+    return true;
+  return false;
```
`return (fileLow.find("index.bdmv") != std::string::npos));`

ksooo replied:
The implementation is not correct. Same problem as `IsDVDFile`.

78andyp replied:
Done

ksooo commented on xbmc/utils/URIUtils.h:
```diff
@@ -135,6 +135,8 @@ class URIUtils
   static bool IsDAV(const std::string& strFile);
   static bool IsDOSPath(const std::string &path);
   static bool IsDVD(const std::string& strFile);
+  static bool IsDVDFile(const std::string& strFile);
+  static bool IsBDFile(const std::string& strFile);
```
Remove `str` prefix from param names. Coding guidelines.

78andyp replied:
Done

ksooo reviewed (commented):
Not a full review. Did not look at the `CFileItem` changes. TL;DR ;-)

CastagnaIT commented:
@78andyp i suggest you to read also [contributing guidelines](https://github.com/xbmc/xbmc/blob/master/docs/CONTRIBUTING.md#pull-request-submission-guidelines)
PR title and commit title must be addressed as explained on the rules

78andyp commented:
> Not a full review. Did not look at the `CFileItem` changes. DL;DR ;-)

At least it works now ;-)

ksooo commented:
> At least it works now ;-)

Yes, but this alone is not a sufficient quality metrics.

ksooo commented on xbmc/FileItem.cpp:
```diff
@@ -2870,289 +2873,225 @@ void CFileItemList::Stack(bool stackFiles /* = true */)
-        item1->m_dwSize = size;
-        break;
+        // Update item
+        CFileItemPtr stackItem{Get(stack[0])};
```
While you are at it, and we want to get rid of `CFileItemPtr`, can you change this to `const auto stackItem{std::make_shared<CFileItem>(Get(stack[0])}; `?

78andyp replied:
No, I think that's an oversight. Thanks.

ksooo commented on xbmc/application/ApplicationStackHelper.h:
```diff
@@ -47,7 +47,7 @@ class CApplicationStackHelper : public IApplicationComponent
   /*!
   \brief Returns true if Application is currently playing an ISO stack
   */
-  bool IsPlayingISOStack() const;
+  bool IsPlayingDiscStack() const;
```
As you changed the name of the method, you should also adjust the documentation.

78andyp replied:
Done.

78andyp replied:
Done

ksooo commented on xbmc/FileItem.cpp:
```diff
@@ -2854,6 +2854,9 @@ void CFileItemList::RemoveExtensions()
+static constexpr int FOLDER_CANDIDATE = 1;
+static constexpr int FILE_CANDIDATE = 2;
+
 void CFileItemList::Stack(bool stackFiles /* = true */)
```
The parameter is not used anymore it seams and thus can be removed.

78andyp replied:
Done

78andyp replied:
Done

ksooo commented on xbmc/FileItem.cpp:
```diff
@@ -2870,289 +2873,225 @@ void CFileItemList::Stack(bool stackFiles /* = true */)
 
-  StackFolders();
-
-  if (stackFiles)
-    StackFiles();
```
Before this PR, stacking of files was done only conditionally, now always. Have you thought about possible performance problems? Just asking.

78andyp replied:
As far as I can see stackFiles was always true - every call was just Stack() so defaulted to true.

ksooo commented on xbmc/FileItem.cpp:
```diff
@@ -2870,289 +2873,225 @@ void CFileItemList::Stack(bool stackFiles /* = true */)
-            item->SetLabelPreformatted(true);
-            m_sortDescription.sortBy = SortByNone; /* sorting is now broken */
+            // Get components of file name
+            const std::string fileTitle = regExp.GetMatch(1);
+            const std::string fileVolume = regExp.GetMatch(2);
```
You don't really need these values as variables, as you need the data only once.

78andyp replied:
It was just to make it more readable. I'll put a comment in instead.

ksooo commented on xbmc/FileItem.cpp:
```diff
@@ -2870,289 +2873,225 @@ void CFileItemList::Stack(bool stackFiles /* = true */)
+    std::sort(stackCandidates.begin(), stackCandidates.end(),
+              [](const auto& s, const auto& t)
+              {
+                if (std::get<0>(s) == std::get<0>(t)) // type (file/folder)
+                  if (std::get<1>(s) == std::get<1>(t)) // title
```
Could you add `{` to make the code a bit more readable?

78andyp replied:
Done

ksooo commented on xbmc/FileItem.cpp:
```diff
@@ -2870,289 +2873,225 @@ void CFileItemList::Stack(bool stackFiles /* = true */)
-                ++expr;
-                break;
-              }
+              CFileItemPtr item{Get(index)};
+              item->SetPath(playPath); // updated path (for DVD/Bluray files)
+              item->m_bIsFolder = false;
+              item->SetLabel2("");
```
This could be rewritten to use a bit more modern C++, and (more important) we want to avoid setting item's path after construction, because this semantically is a change of identity, which strictly speaking must not happen.


```c++
const auto item{std::make_shared<CFileItem>(playPath, false)};
item->SetLabel(Get(index));
```

No need to set an empty label2, btw, as this is the default.

78andyp replied:
Done

78andyp replied:
Needs to be `const std::shared_ptr<CFileItem> item{Get(index)};`

ksooo commented on xbmc/FileItem.cpp:
```diff
@@ -2870,289 +2873,225 @@ void CFileItemList::Stack(bool stackFiles /* = true */)
+      Remove(i);
+
+    // Update non-stack items
+    for (const auto& item : m_items)
+    {
+      const std::string playPath = item->GetOpticalMediaPath();
+      if (!item->IsStack() && !playPath.empty())
+      {
+        item->SetPath(playPath); // updated path (for DVD/Bluray files)
+        item->m_bIsFolder = false;
+        item->SetLabel2("");
+        item->SetLabelPreformatted(true);
+        m_sortDescription.sortBy = SortByNone; // sorting is now broken
```
Not sure whether rewriting m_items' item paths does have bad side effects.

78andyp replied:
This was in original code. In fairness it's completely unrelated to stacks. I'll have a look and see if I can move it before - as SetPath is called in a couple of places for this reason (one for stacks, one for those not stacked).

ksooo commented on xbmc/FileItem.cpp:
```diff
@@ -2870,289 +2873,225 @@ void CFileItemList::Stack(bool stackFiles /* = true */)
+    }
+
+    // Delete unneeded items
+    std::sort(deleteItems.begin(), deleteItems.end(),
+              [](const int i, const int j) { return i > j; });
+    for (int i : deleteItems)
+      Remove(i);
```
Why do you need to sort the items before you remove them?

78andyp replied:
Have to remove highest to lowest otherwise position in array will move.

ksooo commented on xbmc/FileItem.cpp:
```diff
@@ -2870,289 +2873,225 @@ void CFileItemList::Stack(bool stackFiles /* = true */)
-    if (URIUtils::HasEncodedFilename(CURL(filePath)))
-      file1 = CURL::Decode(file1);
+    // Count stack candidates
+    std::map<std::pair<int, std::string>, int> count;
```
Can we find a more suitable name for the is map? By naming it after what it is containing maybe?

78andyp replied:
Done

78andyp commented:
Hi @KarellenX 
If I was to ask for a test build - would you be able to try this out for me??

KarellenX commented:
I don't use stacks, but go ahead. I can whip up some testing for it.

78andyp commented:
> I don't use stacks, but go ahead. I can whip up some testing for it.

Thanks. Try https://file.io/r3JalW6Ma5z6 - should be a release build this time.

KarellenX commented:
@78andyp 

I have tested the test build with the following setup:

1. Two single file *.m2ts files named _Ender's Game (2013) part1_ and _Ender's Game (2013) part2_
2. Two bluray folders that contain the single movie, no other extras. Named _Hateful Eight The (2015) disc1_ and _... disc2_
3. Two iso files that contain the full dvd content, named _Martian (2015) disc1_ and _... disc2_
4. Two dvd folders that contain the single movie, no extras. Named _V for Vendetta (2005) disc1_ and _... disc2_

All 4 movies were correctly identified as stacked movies after scraping...

1. `VideoInfoScanner: Extracted filestream details from video file stack://C:\Kodi\Test- Movie Stacking\Ender's Game (2013)\Ender's Game (2013) part1.m2ts , C:\Kodi\Test- Movie Stacking\Ender's Game (2013)\Ender's Game (2013) part2.m2ts`
2. `VideoInfoScanner: Extracted filestream details from video file stack://C:\Kodi\Test- Movie Stacking\Hateful Eight The (2015) disc1\BDMV\index.bdmv , C:\Kodi\Test- Movie Stacking\Hateful Eight The (2015) disc2\BDMV\index.bdmv`
3. `VideoInfoScanner: Extracted filestream details from video file stack://C:\Kodi\Test- Movie Stacking\Martian (2015) disc1\Natural Wonders of the World 10.iso , C:\Kodi\Test- Movie Stacking\Martian (2015) disc2\Natural Wonders of the World 10.iso`
4. `VideoInfoScanner: Extracted filestream details from video file stack://C:\Kodi\Test- Movie Stacking\V for Vendetta (2005) disc1\VIDEO_TS\VIDEO_TS.IFO , C:\Kodi\Test- Movie Stacking\V for Vendetta (2005) disc2\VIDEO_TS\VIDEO_TS.IFO`

All four movies played back correctly and continued onto the second part/disc at the end of the first part/disc.

1. m2ts - files smoothly transitioned into second file at end of the first file
2. Bluray - At the end of the first file, the Kodi bluray menu appeared and asked to select the item to playback. Item selected and playback continued
3. iso - At the end of the first iso, the dvd menu appeared for the second iso. Item selected and playback continued
4. DVD - At the end of the first file, a dialogue popped up asking to play the next file in the stack. I selected "continue" and playback continued. See image below

As I have only used Stacked files once with single file, I don't know if the menus and dialogue boxes are normal and expected. I guess they are, otherwise how else is the player to know which item to play.

**All seems to be working correctly.**

Image of dialogue from DVD stack playback. Should this also be seen for Bluray and ISO playback?
![stack](https://github.com/xbmc/xbmc/assets/44626605/f8c3f581-0ab0-43ce-b1bf-1367b9572200)

78andyp commented:
> @78andyp
> 
> I have tested the test build with the following setup:
> 
> 1. Two single file *.m2ts files named _Ender's Game (2013) part1_ and _Ender's Game (2013) part2_
> 2. Two bluray folders that contain the single movie, no other extras. Named _Hateful Eight The (2015) disc1_ and _... disc2_
> 3. Two iso files that contain the full dvd content, named _Martian (2015) disc1_ and _... disc2_
> 4. Two dvd folders that contain the single movie, no extras. Named _V for Vendetta (2005) disc1_ and _... disc2_
> 
> All 4 movies were correctly identified as stacked movies after scraping...
> 
> 1. `VideoInfoScanner: Extracted filestream details from video file stack://C:\Kodi\Test- Movie Stacking\Ender's Game (2013)\Ender's Game (2013) part1.m2ts , C:\Kodi\Test- Movie Stacking\Ender's Game (2013)\Ender's Game (2013) part2.m2ts`
> 2. `VideoInfoScanner: Extracted filestream details from video file stack://C:\Kodi\Test- Movie Stacking\Hateful Eight The (2015) disc1\BDMV\index.bdmv , C:\Kodi\Test- Movie Stacking\Hateful Eight The (2015) disc2\BDMV\index.bdmv`
> 3. `VideoInfoScanner: Extracted filestream details from video file stack://C:\Kodi\Test- Movie Stacking\Martian (2015) disc1\Natural Wonders of the World 10.iso , C:\Kodi\Test- Movie Stacking\Martian (2015) disc2\Natural Wonders of the World 10.iso`
> 4. `VideoInfoScanner: Extracted filestream details from video file stack://C:\Kodi\Test- Movie Stacking\V for Vendetta (2005) disc1\VIDEO_TS\VIDEO_TS.IFO , C:\Kodi\Test- Movie Stacking\V for Vendetta (2005) disc2\VIDEO_TS\VIDEO_TS.IFO`
> 
> All four movies played back correctly and continued onto the second part/disc at the end of the first part/disc.
> 
> 1. m2ts - files smoothly transitioned into second file at end of the first file
> 2. Bluray - At the end of the first file, the Kodi bluray menu appeared and asked to select the item to playback. Item selected and playback continued
> 3. iso - At the end of the first iso, the dvd menu appeared for the second iso. Item selected and playback continued
> 4. DVD - At the end of the first file, a dialogue popped up asking to play the next file in the stack. I selected "continue" and playback continued. See image below
> 
> Here is the log... https://paste.kodi.tv/yaputokeva.kodi
> 
> As I have only used Stacked files once with single file, I don't know if the menus and dialogue boxes are normal and expected. I guess they are, otherwise how else is the player to know which item to play.
> 
> **All seems to be working correctly.**
> 
> Image of dialogue from DVD stack playback. Should this also be seen for Bluray and ISO playback? ![stack](https://private-user-images.githubusercontent.com/44626605/321071949-f8c3f581-0ab0-43ce-b1bf-1367b9572200.jpg?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTI3OTEyNzAsIm5iZiI6MTcxMjc5MDk3MCwicGF0aCI6Ii80NDYyNjYwNS8zMjEwNzE5NDktZjhjM2Y1ODEtMGFiMC00M2NlLWIxYmYtMTM2N2I5NTcyMjAwLmpwZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA0MTAlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNDEwVDIzMTYxMFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWJhYzAwN2RiYmU3MmExNThhNzAwMDg1YmNlMzJlMzAzODgzNmY3NDc2YmI5YTJjMmE0ODZmZWIxYTMzNzE1YmImWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.ty_ke5YWxEeTV0A565tuydoUl6GUW-NhlwIURzq3kcg)

Thank you so much @KarellenX, this is really helpful confirmation.

The continue dialog should only appear if you pressed stop to exit a DVD or bluray you played through the menu (and therefore goes back to/would go back to the menu when the film has ended). In this case, as Kodi doesn't 'end' playback automatically - you have to stop it - there is no way to know if you stopped because you had finished or had stopped and want to continue onto the next part. So you would see it (currently) for DVD playback (files or ISO) but not for bluray selected through simple menu (where the playlist selected ends automatically).

I have added code in #24304 so that DVDs are treated the same as blurays - ie you can play titles/playlists directly without having to go through the menu. I'm putting finishing touches to it.

Once that's done i need to test this and it with #24720 which makes playlists/titles treated as files in their own right - ie could have multipe entries for the same disc for different movie versions, each episode is treated seperately (see PR description).

KarellenX commented:
Sure, let me know when the test build is available.

neo1973 commented on xbmc/application/Application.cpp:
```diff
@@ -2253,43 +2254,10 @@ bool CApplication::PlayMedia(CFileItem& item, const std::string& player, PLAYLIS
-                            const std::string& player,
-                            bool bRestart /* = false */,
-                            bool forceSelection /* = false */)
+bool CApplication::PlayFile(CFileItem item, const std::string& player, bool bRestart)bool CApplication::PlayFile(CFileItem item,
```
```suggestion
bool CApplication::PlayFile(CFileItem item,
```

78andyp replied:
Thanks. Spotted this and fixed locally. Just adding bluray:// support to folder stacks (so can navigate between items like with files) and will then update all together.

expldec commented:
This is an amazing coincidence. I installed Kodi just last week and ran into the stacking bug as hundreds of my movies (saved as folders containfing a VIDEO_TS DVD file structure) would not scan correctly.

I (a Javascript developer) looked into the kodi source and slowly realized that my failing folders all had "DVD5" or "DVD9" in the folder name, tricking the code into executing the broken stacking logic you fixed.

You have all my respect for tackling code that hasn't been touched in 10+ years. Thank you so much!

I tested your branch locally, @78andyp  and I no longer see the dreaded "Video (2022)" entries in my library.

I used a SFTP network share and scanned a Movie folder inside the share. I used the TMDB Python scraper

However, I do see about 10% of the movies in my move folder don't get scanned into the library. They only show up in this log message:

`2024-05-03 22:34:41.170 T:26876 warning <general>: No information found for item 'stack://sftp://path/to/movie.title.mkv , sftp://path/to/movie.title2.mkv, sftp://path/to/movie.title3.mkv`

I redacted the names (i can provide them privtely if you like), but they are all .mkv files. all totally unrelated movies from one another.

Could it be that file items end up in a stack that `CVideoInfoScanner::RetrieveInfoForMovie` can't handle? I would be happy if only the files would show up somewhere in the Movie library, even with unscraped titles, instead of being completely unreachable.

78andyp commented:
> This is an amazing coincidence. I installed Kodi just last week and ran into the stacking bug as hundreds of my movies (saved as folders containfing a VIDEO_TS DVD file structure) would not scan correctly.
> 
> I (a Javascript developer) looked into the kodi source and slowly realized that my failing folders all had "DVD5" or "DVD9" in the folder name, tricking the code into executing the broken stacking logic you fixed.
> 
> You have all my respect for tackling code that hasn't been touched in 10+ years. Thank you so much!
> 
> I tested your branch locally, @78andyp and I no longer see the dreaded "Video (2022)" entries in my library.
> 
> I used a SFTP network share and scanned a Movie folder inside the share. I used the TMDB Python scraper
> 
> However, I do see about 10% of the movies in my movie folder don't get scanned into the library. They only show up in this log message:
> 
> `2024-05-03 22:34:41.170 T:26876 warning <general>: No information found for item 'stack://sftp://path/to/movie.title.mkv , sftp://path/to/movie.title2.mkv, sftp://path/to/movie.title3.mkv`
> 
> I redacted the names (i can provide them privately if you like), but they are all .mkv files. all totally unrelated movies from one another.
> 
> Could it be that file items end up in a stack that `CVideoInfoScanner::RetrieveInfoForMovie` can't handle? I would be happy if only the files would show up somewhere in the Movie library, even with unscraped titles, instead of being completely unreachable.

Hi
Thanks for testing :)
Please do PM me the names - my userid in the forum is powellandy.

expldec commented:
I investigated this further. I think I found a good clue as to what happens and it's kinda funny.

It's one of the `m_videoStackRegExps`

```m_videoStackRegExps.emplace_back("(.*?)([ ._-]*[a-d])(.*?)(\\.[^.]+)$");```

When run against a folder that contains multiple videos, it will match every filename that contains the letters "a" through "d".
The first capture group ends up being empty for most of the matches.

This means that all matched movies will be put in stackCandidates, and grouped together into a stack with an empty title.

Personally I'm having trouble figuring out the usefulness of this regex. Especially it being there by default. I think it should be removed.

Another good idea would be to not put an item in stackCandidates if the regex matches an empty title.

Thoughts?

78andyp commented:
> I investigated this further. I think I found a good clue as to what happens and it's kinda funny.
> 
> It's one of the `m_videoStackRegExps`
> 
> `m_videoStackRegExps.emplace_back("(.*?)([ ._-]*[a-d])(.*?)(\\.[^.]+)$");`
> 
> When run against a folder that contains multiple videos, it **will match every filename that contains the letters "a" through "d"**. The first capture group ends up being empty for most of the matches.
> 
> This means that all matched movies will be put in `stackCandidates`, and grouped together into a stack with an empty title.
> 
> What's worse, the regex will end up stacking titles that don't belong together, e.g.
> 
> ```
> The.Chronicles.of.Narnia.The Lion,.the.Witch.and.the.Wardrobe.2005.1080p.mkv
> The.Chronicles.of.Narnia.The.Voyage.of.the.Dawn.Treader.2010.1080p.mkv
> The.Chronicles.of.Riddick.2004.1080p.mkv
> The.Captain.mkv
> ```
> 
> Will all be stacked under the title "The"
> 
> Personally I'm having trouble figuring out the usefulness of this regex. Especially it being there by default. I think it should be removed.
> 
> Another good idea would be to not put an item in `stackCandidates` if the regex matches an empty title.
> 
> Thoughts?

Thanks, and thanks for the DM.

I haven't actually looked yet but in my mind I was thinking regex so I'm pretty sure you are right, thanks.

I assume the intent ot that regex is a file name ending in .A or -A etc.. would trigger.

If so, perhaps this regex would be better - `(.+)(?:[ ._-]+)([a-d])(.*?)(\\.[^.]+)$'

expldec commented:
I tested again after changing the regex to your proposed one.

It did work better simply because the (.+) at the start force it to match at least one character.
However it will still stack the examples I gave above under the title "the"

IMO the best course of action is to comment it out and let the few users that will find it useful add it to their advancedSettings.xml.
Looking for "anything followed by a/b/c/d" is far too crude to work in a flat directory structure and will cause the scan to miss A LOT of titles overall.

78andyp commented:
They shouldn't do as `(?:[ ._-]+)([a-d])` means there must be a `.` `_` or `-` followed by a letter `a-d`.
None of the names you give trigger the regex at https://regex101.com/ when `(.+)(?:[ ._-]+)([a-d])(.*?)(\.[^.]+)$` is used.

expldec commented:
> None of the names you give trigger the regex at https://regex101.com/ when `(.+)(?:[ ._-]+)([a-d])(.*?)(\.[^.]+)$` is used.

Hey, thanks for following up :). In my testing they actually do. Keep in mind that `URIUtils::Split(StringUtils::ToLower(item->GetPath()), filePath, file);` is run on the item path so the filenames get converted to lowercase before being tested by the regex. (in the screenshot this is simulated by adding the `i` flag to the regex)

![image](https://github.com/xbmc/xbmc/assets/16882478/594d97e4-3263-4ad4-a64d-6a7b56c89bc7)

78andyp commented:
> > None of the names you give trigger the regex at https://regex101.com/ when `(.+)(?:[ ._-]+)([a-d])(.*?)(\.[^.]+)$` is used.
> 
> Hey, thanks for following up :). In my testing they actually do. Keep in mind that `URIUtils::Split(StringUtils::ToLower(item->GetPath()), filePath, file);` is run on the item path so the filenames get converted to lowercase before being tested by the regex. (in the screenshot this is simulated by adding the `i` flag to the regex)
> 
> ![image](https://private-user-images.githubusercontent.com/16882478/328177159-594d97e4-3263-4ad4-a64d-6a7b56c89bc7.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTUwMzYxMTEsIm5iZiI6MTcxNTAzNTgxMSwicGF0aCI6Ii8xNjg4MjQ3OC8zMjgxNzcxNTktNTk0ZDk3ZTQtMzI2My00YWQ0LWE2NGQtNmE3YjU2Yzg5YmM3LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA1MDYlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNTA2VDIyNTAxMVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWE3ZmUxZGM5NGZjMDAzNzY2YmJkYjU1MTllNTc3NWFlYWVjNDVlNzU3OWRmYmY5ZWMxODVlMjg3YTA4MTI2NmMmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.Is29BgFSeNLQ_FHKl75ZX-yXEvx9YgVat0g3VHMr-r0)

You are quite correct.

I would still like the thoughts of those with greater experience before we remove it.

How about `(.+)(?:[ ._-]+)([a-d])()(\.[^.]+)$` - makes the `a-d` only match if at the end of the filename before the extension.

expldec commented:
> How about `(.+)(?:[ ._-]+)([a-d])()(\.[^.]+)$` - makes the `a-d` only match if at the end of the filename before the extension.

Definitely better, I think this is restrictive enough to be included by default. People that need it looser are always free to include a custom one in their advancedsettings.xml :)

78andyp commented:
> > How about `(.+)(?:[ ._-]+)([a-d])()(\.[^.]+)$` - makes the `a-d` only match if at the end of the filename before the extension.
> 
> Definitely better, I think this is restrictive enough to be included by default. People that need it looser are always free to include a custom one in their advancedsettings.xml :)

Thanks.

I've tweaked it a little as there is actually a test case specifically for this that was failing - see `\xbmc\video\test\testdata\moviestack_ab\Movie-(2001)` where it requires `Movie-(2001)A.mp4` and `Movie-(2001)B.mp4` to be made into a stack.

`^(.+)((?:[ ._-]|(?<=\)))[a-z])()(\\.[^.]+)$`

is what I've settled on - does this work for you??

expldec commented:
Your last regex seems malformed. the parentheses don't match up. What is the reasoning behind that positive lookbehind?

I think this one should pass the test: `^(.+)(?:[ ._\)-]+)([a-d])()(\.[^.]+)$` 

It just add a closing parenthesis to the accepted delimiters in front of a-d (or a-z if you prefer)

78andyp commented:
> Your last regex seems malformed. the parentheses don't match up. What is the reasoning behind that positive lookbehind?
> 
> I think this one should pass the test: `^(.+)(?:[ ._\)-]+)([a-d])()(\.[^.]+)$`
> 
> It just add a closing parenthesis to the accepted delimiters in front of a-d (or a-z if you prefer)

Sorry copy paste error:

![image](https://github.com/xbmc/xbmc/assets/99039295/253e6560-6798-421d-b703-15ba1a16f9b1)

It's so a `)` can be one of the characters that matches prior to the letter but it's preserved in the movie name.

expldec commented:
> Possibly copy paste error

I tried the one from your screenshot. that works! I have no objection to that :)

78andyp commented:
@ksooo, @CrystalP  - would you be able to review the code, if you have a moment?
@KarellenX - could you repeat your testing, if you have the time?? I've not changed the stack creation but i've tinkered with other bits now bluray:// support is in - looking at playing/stopping/resuming/moving back/fowards between parts of bluray stacks.
Thanks all.

KarellenX commented:
Win64 test build here... https://mirrors.kodi.tv/test-builds/windows/win64/KodiSetup-20240518-8fe8f626-stacks-x64.exe

@78andyp Sure, I'll test today or tomorrow.

78andyp committed:
[Video][Stacks] Fix folder stacks.

KarellenX commented:
Why did you close @78andyp ?
Are you opening a new PR?

78andyp commented:
> Why did you close @78andyp ? Are you opening a new PR?

Hi
Apologies, I should have made clearer. This is now superceded by #25281.
This was always going to be (slightly) incomplete - in that it could dynamically generate bluray folder stacks (ie.. the first time you watch the stack it would ask for each bluray title as you got to the disc (folder/iso), and then ultimately the completed stack would play like a file stack - seamlessly moving between all parts) but it needed to be merged with #24997 to allow the same for dvd folders/isos. As there was a delay in getting it in I thought I might as well add all the functionality at once - so now there is complete folder stack functionality for blurays and dvds.
Thanks
Andy

KarellenX commented:
Ok thanks. I completely missed your new PR.
